File: backend/src/voice/protocol.py
"""WebSocket message protocol definitions for voice tutor."""

from enum import Enum
from typing import Any, Dict, Literal, Optional
from pydantic import BaseModel

from voice.config import DEFAULT_VOICE_ID


# --- Session States ---

class SessionState(str, Enum):
    IDLE = "idle"
    LISTENING = "listening"
    PROCESSING = "processing"
    SPEAKING = "speaking"


# --- Frontend -> Backend Messages ---

class StartSessionMsg(BaseModel):
    type: Literal["start_session"] = "start_session"
    user_id: str
    session_id: str
    voice_id: str = DEFAULT_VOICE_ID
    config: Dict[str, Any] = {}


class AudioDataMsg(BaseModel):
    type: Literal["audio_data"] = "audio_data"
    audio: str  # Base64 encoded audio
    format: str = "pcm"  # "pcm" or "wav"


class StopAudioMsg(BaseModel):
    type: Literal["stop_audio"] = "stop_audio"


class InterruptMsg(BaseModel):
    type: Literal["interrupt"] = "interrupt"


class PlaybackFinishedMsg(BaseModel):
    type: Literal["playback_finished"] = "playback_finished"


class EndSessionMsg(BaseModel):
    type: Literal["end_session"] = "end_session"


class SwitchVoiceMsg(BaseModel):
    type: Literal["switch_voice"] = "switch_voice"
    voice_id: str


class UpdateConfigMsg(BaseModel):
    type: Literal["update_config"] = "update_config"
    config: Dict[str, Any] = {}


# --- Backend -> Frontend Messages ---

class SessionStartedMsg(BaseModel):
    type: str = "session_started"
    session_id: str


class STTPartialMsg(BaseModel):
    type: str = "stt_partial"
    text: str
    is_final: bool = False


class STTFinalMsg(BaseModel):
    type: str = "stt_final"
    text: str


class AgentThinkingMsg(BaseModel):
    type: str = "agent_thinking"


class AgentTextMsg(BaseModel):
    type: str = "agent_text"
    text: str
    is_final: bool = False


class AudioChunkMsg(BaseModel):
    type: str = "audio_chunk"
    audio: str  # Base64 encoded audio
    is_last: bool = False


class AudioStreamEndMsg(BaseModel):
    """Sent when all TTS audio chunks have been delivered."""
    type: str = "audio_stream_end"


class SessionEndedMsg(BaseModel):
    type: str = "session_ended"
    reason: str = "normal"


class ErrorMsg(BaseModel):
    type: str = "error"
    code: str
    message: str


# --- Helper ---

def parse_frontend_msg(data: dict) -> BaseModel:
    """Parse incoming WebSocket message into the appropriate model."""
    msg_type = data.get("type")
    parsers = {
        "start_session": StartSessionMsg,
        "audio_data": AudioDataMsg,
        "stop_audio": StopAudioMsg,
        "interrupt": InterruptMsg,
        "playback_finished": PlaybackFinishedMsg,
        "end_session": EndSessionMsg,
        "switch_voice": SwitchVoiceMsg,
        "update_config": UpdateConfigMsg,
    }
    parser = parsers.get(msg_type)
    if parser is None:
        raise ValueError(f"Unknown message type: {msg_type}")
    return parser(**data)

File: frontend/src/utils/protocol.js
// 前端发送的消息类型
export const CLIENT_MESSAGES = {
  START_SESSION: 'start_session',
  AUDIO_DATA: 'audio_data',
  STOP_AUDIO: 'stop_audio',
  INTERRUPT: 'interrupt',
  PLAYBACK_FINISHED: 'playback_finished',
  END_SESSION: 'end_session',
  SWITCH_VOICE: 'switch_voice',
  UPDATE_CONFIG: 'update_config',
};

// 后端返回的消息类型
export const SERVER_MESSAGES = {
  SESSION_STARTED: 'session_started',
  STT_PARTIAL: 'stt_partial',
  STT_FINAL: 'stt_final',
  AGENT_THINKING: 'agent_thinking',
  AGENT_TEXT: 'agent_text',
  AUDIO_CHUNK: 'audio_chunk',
  AUDIO_STREAM_END: 'audio_stream_end',
  STOP_CAPTURE: 'stop_capture',
  START_CAPTURE: 'start_capture',
  SESSION_ENDED: 'session_ended',
  ERROR: 'error',
};

// 生成会话ID
export function generateSessionId() {
  return `sess-voice-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

// 创建开始会话消息
export function createStartSessionMessage(userId, voiceId, chatHistory = []) {
  const conversationHistory = chatHistory.slice(-10).map((item) => ({
    role: item.role === 'ai' ? 'assistant' : 'user',
    content: item.text,
  }));

  return {
    type: CLIENT_MESSAGES.START_SESSION,
    user_id: userId,
    session_id: generateSessionId(),
    voice_id: voiceId,
    config: {
      conversation_history: conversationHistory,
    },
  };
}

// 创建音频数据消息
export function createAudioDataMessage(base64Audio) {
  return {
    type: CLIENT_MESSAGES.AUDIO_DATA,
    audio: base64Audio,
    format: 'pcm',
  };
}

// 创建停止本轮语音输入消息
export function createStopAudioMessage() {
  return {
    type: CLIENT_MESSAGES.STOP_AUDIO,
  };
}

// 创建打断消息
export function createInterruptMessage() {
  return {
    type: CLIENT_MESSAGES.INTERRUPT,
  };
}

// 创建播放完成消息
export function createPlaybackFinishedMessage() {
  return {
    type: CLIENT_MESSAGES.PLAYBACK_FINISHED,
  };
}

// 创建结束会话消息
export function createEndSessionMessage() {
  return {
    type: CLIENT_MESSAGES.END_SESSION,
  };
}

// 创建切换音色消息
export function createSwitchVoiceMessage(voiceId) {
  return {
    type: CLIENT_MESSAGES.SWITCH_VOICE,
    voice_id: voiceId,
  };
}
